diff --git a/build/soong/service_fuzzer_bindings.go b/build/soong/service_fuzzer_bindings.go
index 63a07e9458dd5bd6860b934831a86cd65320a647..27c6eb1662f34a512d0d491abb5616fb43e542b8 100644
--- a/build/soong/service_fuzzer_bindings.go
+++ b/build/soong/service_fuzzer_bindings.go
@@ -437,6 +437,7 @@ var (
 		"tv_tuner_resource_mgr":        EXCEPTION_NO_FUZZER,
 		"uce":                          EXCEPTION_NO_FUZZER,
 		"uimode":                       EXCEPTION_NO_FUZZER,
+		"ufam_security":                EXCEPTION_NO_FUZZER,
 		"updatelock":                   EXCEPTION_NO_FUZZER,
 		"uri_grants":                   EXCEPTION_NO_FUZZER,
 		"usagestats":                   EXCEPTION_NO_FUZZER,
diff --git a/prebuilts/api/34.0/private/service_contexts b/prebuilts/api/34.0/private/service_contexts
index 553a71df82ec273a580985f6e33547dbc17ca392..2b0e595861648c5b16aee08cd6b391a8ef28a1d5 100644
--- a/prebuilts/api/34.0/private/service_contexts
+++ b/prebuilts/api/34.0/private/service_contexts
@@ -414,6 +414,7 @@ tv_input                                  u:object_r:tv_input_service:s0
 tv_tuner_resource_mgr                     u:object_r:tv_tuner_resource_mgr_service:s0
 uce                                       u:object_r:uce_service:s0
 uimode                                    u:object_r:uimode_service:s0
+ufam_security                             u:object_r:device_policy_service:s0
 updatelock                                u:object_r:updatelock_service:s0
 uri_grants                                u:object_r:uri_grants_service:s0
 usagestats                                u:object_r:usagestats_service:s0
diff --git a/private/service_contexts b/private/service_contexts
index 553a71df82ec273a580985f6e33547dbc17ca392..2b0e595861648c5b16aee08cd6b391a8ef28a1d5 100644
--- a/private/service_contexts
+++ b/private/service_contexts
@@ -414,6 +414,7 @@ tv_input                                  u:object_r:tv_input_service:s0
 tv_tuner_resource_mgr                     u:object_r:tv_tuner_resource_mgr_service:s0
 uce                                       u:object_r:uce_service:s0
 uimode                                    u:object_r:uimode_service:s0
+ufam_security                             u:object_r:device_policy_service:s0
 updatelock                                u:object_r:updatelock_service:s0
 uri_grants                                u:object_r:uri_grants_service:s0
 usagestats                                u:object_r:usagestats_service:s0
